File: src/main/kotlin/cn/nukkit/block/BlockBed.java
package cn.nukkit.block;

import cn.nukkit.Player;
import cn.nukkit.blockentity.BlockEntity;
import cn.nukkit.blockentity.BlockEntityBed;
import cn.nukkit.entity.Entity;
import cn.nukkit.entity.effect.EffectType;
import cn.nukkit.event.block.BlockExplosionPrimeEvent;
import cn.nukkit.item.Item;
import cn.nukkit.item.ItemBed;
import cn.nukkit.lang.TranslationContainer;
import cn.nukkit.level.Explosion;
import cn.nukkit.level.GameRule;
import cn.nukkit.level.Level;
import cn.nukkit.level.Transform;
import cn.nukkit.math.AxisAlignedBB;
import cn.nukkit.math.BlockFace;
import cn.nukkit.math.SimpleAxisAlignedBB;
import cn.nukkit.nbt.tag.CompoundTag;
import cn.nukkit.network.protocol.types.SpawnPointType;
import cn.nukkit.utils.DyeColor;
import cn.nukkit.utils.Faceable;
import cn.nukkit.utils.TextFormat;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

import javax.annotation.Nullable;

import static cn.nukkit.block.property.CommonBlockProperties.DIRECTION;
import static cn.nukkit.block.property.CommonBlockProperties.HEAD_PIECE_BIT;
import static cn.nukkit.block.property.CommonBlockProperties.OCCUPIED_BIT;

/**
 * @author MagicDroidX (Nukkit Project)
 */

@Slf4j
public class BlockBed extends BlockTransparent implements Faceable, BlockEntityHolder<BlockEntityBed> {
    public static final BlockProperties PROPERTIES = new BlockProperties(BED, DIRECTION, HEAD_PIECE_BIT, OCCUPIED_BIT);

    @Override
    @NotNull public BlockProperties getProperties() {
        return PROPERTIES;
    }

    public BlockBed() {
        this(PROPERTIES.getDefaultState());
    }

    public BlockBed(BlockState blockstate) {
        super(blockstate);
    }

    @Override
    @NotNull public Class<? extends BlockEntityBed> getBlockEntityClass() {
        return BlockEntityBed.class;
    }

    @Override
    @NotNull public String getBlockEntityType() {
        return BlockEntity.BED;
    }

    @Override
    public boolean canBeActivated() {
        return true;
    }

    @Override
    public double getResistance() {
        return 1;
    }

    @Override
    public double getHardness() {
        return 0.2;
    }

    @Override
    public String getName() {
        return this.getDyeColor().getName() + " Bed Block";
    }

    @Override
    public double getMaxY() {
        return this.position.up + 0.5625;
    }

    @Override
    public int getWaterloggingLevel() {
        return 1;
    }


    @Override
    public boolean onActivate(@NotNull Item item, Player player, BlockFace blockFace, float fx, float fy, float fz) {
        if(isNotActivate(player)) return false;
        BlockFace dir = getBlockFace();

        boolean shouldExplode = this.level.getDimension() != Level.DIMENSION_OVERWORLD;
        boolean willExplode = shouldExplode && this.level.gameRules.getBoolean(GameRule.TNT_EXPLODES);

        Block head;
        if (isHeadPiece()) {
            head = this;
        } else {
            head = getSide(dir);
            if (head.getId() != getId() || !((BlockBed) head).isHeadPiece() || !((BlockBed) head).getBlockFace().equals(dir)) {
                if (player != null && !willExplode) {
                    player.sendMessage(new TranslationContainer(TextFormat.GRAY + "%tile.bed.notValid"));
                }

                if (!shouldExplode) {
                    return true;
                }
            }
        }

        BlockFace footPart = dir.getOpposite();

        if (shouldExplode) {
            if (!willExplode) {
                return true;
            }

            BlockExplosionPrimeEvent event = new BlockExplosionPrimeEvent(this, player, 5);
            event.setIncendiary(true);
            if (event.isCancelled()) {
                return true;
            }

            level.setBlock(this.position, get(AIR), false, false);
            onBreak(Item.AIR);
            level.updateAround(this.position);

            Explosion explosion = new Explosion(this, event.force, this);
            explosion.fireChance = event.fireChance;
            if (event.isBlockBreaking()) {
                explosion.explodeA();
            }
            explosion.explodeB();
            return true;
        }

        if (player == null || !player.hasEffect(EffectType.CONDUIT_POWER) && getLevelBlockAtLayer(1) instanceof BlockFlowingWater) {
            return true;
        }

        AxisAlignedBB accessArea = new SimpleAxisAlignedBB(head.position.south - 2, head.position.up - 5.5, head.position.west - 2, head.position.south + 3, head.position.up + 2.5, head.position.west + 3)
                .addCoord(footPart.getXOffset(), 0, footPart.getZOffset());

        if (!accessArea.isVectorInside(player.position)) {
            player.sendMessage(new TranslationContainer(TextFormat.GRAY + "%tile.bed.tooFar"));
            return true;
        }

        Transform spawn = Transform.fromObject(this.position.add(0.5, 0.5, 0.5), player.level, player.rotation.yaw, player.rotation.pitch);
        if (!player.getSpawn().equals(spawn)) {
            player.setSpawn(this, SpawnPointType.BLOCK);
        }
        player.sendMessage(new TranslationContainer(TextFormat.GRAY + "%tile.bed.respawnSet"));

        int time = this.level.getTime() % Level.TIME_FULL;

        boolean isNight = (time >= Level.TIME_NIGHT && time < Level.TIME_SUNRISE);

        if (!isNight) {
            player.sendMessage(new TranslationContainer(TextFormat.GRAY + "%tile.bed.noSleep"));
            return true;
        }

        if (!player.isCreative()) {
            AxisAlignedBB checkMonsterArea = new SimpleAxisAlignedBB(head.position.south - 8, head.position.up - 6.5, head.position.west - 8, head.position.south + 9, head.position.up + 5.5, head.position.west + 9)
                    .addCoord(footPart.getXOffset(), 0, footPart.getZOffset());

            for (Entity entity : this.level.getCollidingEntities(checkMonsterArea)) {
                if (!entity.isClosed() && entity.isPreventingSleep(player)) {
                    player.sendTranslation(TextFormat.GRAY + "%tile.bed.notSafe");
                    return true;
                }
                // TODO: Check Chicken Jockey, Spider Jockey
            }
        }

        if (!player.sleepOn(head.position)) {
            player.sendMessage(new TranslationContainer(TextFormat.GRAY + "%tile.bed.occupied"));
        }

        return true;
    }

    @Override
    public boolean place(@NotNull Item item, @NotNull Block block, @NotNull Block target, @NotNull BlockFace face, double fx, double fy, double fz, @Nullable Player player) {
        Block down = this.down();
        if (!(BlockLever.isSupportValid(down, BlockFace.UP) || down instanceof BlockCauldron)) {
            return false;
        }

        BlockFace direction = player == null ? BlockFace.NORTH : player.getDirection();
        Block next = this.getSide(direction);
        Block downNext = next.down();

        if (!next.canBeReplaced() || !(BlockLever.isSupportValid(downNext, BlockFace.UP) || downNext instanceof BlockCauldron)) {
            return false;
        }

        Block thisLayer0 = level.getBlock(this.position, 0);
        Block thisLayer1 = level.getBlock(this.position, 1);
        Block nextLayer0 = level.getBlock(next.position, 0);
        Block nextLayer1 = level.getBlock(next.position, 1);

        setBlockFace(direction);

        level.setBlock(block.position, this, true, true);
        if (next instanceof BlockLiquid && ((BlockLiquid) next).usesWaterLogging()) {
            level.setBlock(next.position, 1, next, true, false);
        }

        BlockBed head = (BlockBed) clone();
        head.setHeadPiece(true);
        level.setBlock(next.position, head, true, true);

        BlockEntityBed thisBed = null;
        try {
            thisBed = createBlockEntity(new CompoundTag().putByte("color", item.getDamage()));
            BlockEntityHolder<?> nextBlock = (BlockEntityHolder<?>) next.getLevelBlock();
            nextBlock.createBlockEntity(new CompoundTag().putByte("color", item.getDamage()));
        } catch (Exception e) {
            log.warn("Failed to create the block entity {} at {} and {}", getBlockEntityType(), getLocator(), next.getLocator(), e);
            if (thisBed != null) {
                thisBed.close();
            }
            level.setBlock(thisLayer0.position, 0, thisLayer0, true);
            level.setBlock(thisLayer1.position, 1, thisLayer1, true);
            level.setBlock(nextLayer0.position, 0, nextLayer0, true);
            level.setBlock(nextLayer1.position, 1, nextLayer1, true);
            return false;
        }
        return true;
    }

    @Override
    public boolean onBreak(Item item) {
        BlockFace direction = getBlockFace();
        if (isHeadPiece()) { //This is the Top part of bed
            Block other = getSide(direction.getOpposite());
            if (other.getId().equals(getId()) && !((BlockBed) other).isHeadPiece() && direction.equals(((BlockBed) other).getBlockFace())) {
                level.setBlock(other.position, Block.get(BlockID.AIR), true, true);
            }
        } else { //Bottom Part of Bed
            Block other = getSide(direction);
            if (other.getId().equals(getId()) && ((BlockBed) other).isHeadPiece() && direction.equals(((BlockBed) other).getBlockFace())) {
                level.setBlock(other.position, Block.get(BlockID.AIR), true, true);
            }
        }

        level.setBlock(this.position, Block.get(BlockID.AIR), true, false); // Do not update both parts to prevent duplication bug if there is two fallable blocks top of the bed

        return true;
    }

    @Override
    public Item toItem() {
        return new ItemBed(this.getDyeColor().getWoolData());
    }

    public DyeColor getDyeColor() {
        if (this.level != null) {
            BlockEntityBed blockEntity = getBlockEntity();

            if (blockEntity != null) {
                return blockEntity.getDyeColor();
            }
        }

        return DyeColor.WHITE;
    }

    @Override
    public BlockFace getBlockFace() {
        return BlockFace.fromHorizontalIndex(getPropertyValue(DIRECTION));
    }

    @Override
    public void setBlockFace(BlockFace face) {
        setPropertyValue(DIRECTION, face.horizontalIndex);
    }

    public boolean isHeadPiece() {
        return getPropertyValue(HEAD_PIECE_BIT);
    }

    public void setHeadPiece(boolean headPiece) {
        setPropertyValue(HEAD_PIECE_BIT, headPiece);
    }

    public boolean isOccupied() {
        return getPropertyValue(OCCUPIED_BIT);
    }

    public void setOccupied(boolean occupied) {
        setPropertyValue(OCCUPIED_BIT, occupied);
    }

    @Override
    public boolean breaksWhenMoved() {
        return true;
    }

    @Override
    public boolean sticksToPiston() {
        return false;
    }

    public boolean isBedValid() {
        BlockFace dir = getBlockFace();
        Block head;
        Block foot;
        if (isHeadPiece()) {
            head = this;
            foot = getSide(dir.getOpposite());
        } else {
            head = getSide(dir);
            foot = this;
        }

        return head.getId().equals(foot.getId())
                && ((BlockBed) head).isHeadPiece() && ((BlockBed) head).getBlockFace().equals(dir)
                && !((BlockBed) foot).isHeadPiece() && ((BlockBed) foot).getBlockFace().equals(dir);
    }

    public @Nullable BlockBed getHeadPart() {
        if (isHeadPiece()) {
            return this;
        }
        BlockFace dir = getBlockFace();
        Block head = getSide(dir);
        if (head.getId().equals(getId()) && ((BlockBed) head).isHeadPiece() && ((BlockBed) head).getBlockFace().equals(dir)) {
            return (BlockBed) head;
        }
        return null;
    }

    public @Nullable BlockBed getFootPart() {
        if (!isHeadPiece()) {
            return this;
        }

        BlockFace dir = getBlockFace();
        Block foot = getSide(dir.getOpposite());
        if (foot.getId().equals(getId()) && !((BlockBed) foot).isHeadPiece() && ((BlockBed) foot).getBlockFace().equals(dir)) {
            return (BlockBed) foot;
        }
        return null;
    }
}

File: src/main/kotlin/org/chorus/block/BlockBed.kt
package org.chorus.block

import org.chorus.Player
import org.chorus.block.property.CommonBlockProperties
import org.chorus.block.property.type.BooleanPropertyType
import org.chorus.block.property.type.IntPropertyType
import org.chorus.blockentity.BlockEntityBed
import org.chorus.blockentity.BlockEntityID
import org.chorus.entity.effect.EffectType
import org.chorus.event.block.BlockExplosionPrimeEvent
import org.chorus.item.Item
import org.chorus.item.ItemBed
import org.chorus.lang.TranslationContainer
import org.chorus.level.Explosion
import org.chorus.level.GameRule
import org.chorus.level.Level
import org.chorus.math.BlockFace
import org.chorus.math.BlockFace.Companion.fromHorizontalIndex
import org.chorus.math.SimpleAxisAlignedBB
import org.chorus.nbt.tag.CompoundTag
import org.chorus.network.protocol.types.SpawnPointType
import org.chorus.utils.DyeColor
import org.chorus.utils.Faceable
import org.chorus.utils.Loggable
import org.chorus.utils.TextFormat


class BlockBed @JvmOverloads constructor(blockstate: BlockState = Companion.properties.defaultState) :
    BlockTransparent(blockstate), Faceable, BlockEntityHolder<BlockEntityBed>, Loggable {
    override fun getBlockEntityClass(): Class<out BlockEntityBed> {
        return BlockEntityBed::class.java
    }

    override fun getBlockEntityType(): String {
        return BlockEntityID.BED
    }

    override fun canBeActivated(): Boolean {
        return true
    }

    override val resistance: Double
        get() = 1.0

    override val hardness: Double
        get() = 0.2

    override val name: String
        get() = dyeColor.colorName + " Bed Block"

    override var maxY: Double
        get() = position.y + 0.5625
        set(maxY) {
            super.maxY = maxY
        }

    override val waterloggingLevel: Int
        get() = 1


    override fun onActivate(
        item: Item,
        player: Player?,
        blockFace: BlockFace,
        fx: Float,
        fy: Float,
        fz: Float
    ): Boolean {
        if (isNotActivate(player)) return false
        val dir = blockFace

        val shouldExplode = level.dimension != Level.DIMENSION_OVERWORLD
        val willExplode = shouldExplode && level.gameRules.getBoolean(GameRule.TNT_EXPLODES)

        val head: Block?
        if (isHeadPiece) {
            head = this
        } else {
            head = getSide(dir!!)
            if (head!!.id !== id || !(head as BlockBed).isHeadPiece || (head.blockFace != dir)) {
                if (player != null && !willExplode) {
                    player.sendMessage(TranslationContainer(TextFormat.GRAY.toString() + "%tile.bed.notValid"))
                }

                if (!shouldExplode) {
                    return true
                }
            }
        }

        val footPart = dir!!.getOpposite()

        if (shouldExplode) {
            if (!willExplode) {
                return true
            }

            val event = BlockExplosionPrimeEvent(this, player, 5.0)
            event.isIncendiary = true
            if (event.isCancelled) {
                return true
            }

            level.setBlock(this.position, get(BlockID.AIR), direct = false, update = false)
            onBreak(Item.AIR)
            level.updateAround(this.position)

            val explosion = Explosion(this, event.force, this)
            explosion.fireChance = event.fireChance
            if (event.isBlockBreaking) {
                explosion.explodeA()
            }
            explosion.explodeB()
            return true
        }

        if (player == null || !player.hasEffect(EffectType.CONDUIT_POWER) && getLevelBlockAtLayer(1) is BlockFlowingWater) {
            return true
        }

        val accessArea = SimpleAxisAlignedBB(
            head!!.position.x - 2,
            head.position.y - 5.5,
            head.position.z - 2,
            head.position.x + 3,
            head.position.y + 2.5,
            head.position.z + 3
        )
            .addCoord(footPart!!.xOffset.toDouble(), 0.0, footPart.zOffset.toDouble())

        if (!accessArea.isVectorInside(player.position)) {
            player.sendMessage(TranslationContainer(TextFormat.GRAY.toString() + "%tile.bed.tooFar"))
            return true
        }

        val spawn = fromObject(
            position.add(0.5, 0.5, 0.5),
            player.level!!
        )
        if (player.spawn.first() != spawn) {
            player.setSpawn(this, SpawnPointType.BLOCK)
        }
        player.sendMessage(TranslationContainer(TextFormat.GRAY.toString() + "%tile.bed.respawnSet"))

        val time = level.getTime() % Level.TIME_FULL

        val isNight = (time >= Level.TIME_NIGHT && time < Level.TIME_SUNRISE)

        if (!isNight) {
            player.sendMessage(TranslationContainer(TextFormat.GRAY.toString() + "%tile.bed.noSleep"))
            return true
        }

        if (!player.isCreative) {
            val checkMonsterArea = SimpleAxisAlignedBB(
                head.position.x - 8,
                head.position.y - 6.5,
                head.position.z - 8,
                head.position.x + 9,
                head.position.y + 5.5,
                head.position.z + 9
            )
                .addCoord(footPart.xOffset.toDouble(), 0.0, footPart.zOffset.toDouble())

            for (entity in level.getCollidingEntities(checkMonsterArea)) {
                if (entity != null && !entity.isClosed() && entity.isPreventingSleep(player)) {
                    player.sendTranslation(TextFormat.GRAY.toString() + "%tile.bed.notSafe")
                    return true
                }
                // TODO: Check Chicken Jockey, Spider Jockey
            }
        }

        if (!player.sleepOn(head.position)) {
            player.sendMessage(TranslationContainer(TextFormat.GRAY.toString() + "%tile.bed.occupied"))
        }

        return true
    }

    override fun place(
        item: Item,
        block: Block,
        target: Block,
        face: BlockFace,
        fx: Double,
        fy: Double,
        fz: Double,
        player: Player?
    ): Boolean {
        val down = this.down()
        if (down != null && !(BlockLever.isSupportValid(down, BlockFace.UP) || down is BlockCauldron)) {
            return false
        }

        val direction = if (player == null) BlockFace.NORTH else player.getDirection()
        val next = this.getSide(direction!!)
        val downNext = next!!.down()

        if (downNext != null && (!next.canBeReplaced() || !(BlockLever.isSupportValid(
                downNext,
                BlockFace.UP
            ) || downNext is BlockCauldron))
        ) {
            return false
        }

        val thisLayer0 = level.getBlock(this.position, 0)
        val thisLayer1 = level.getBlock(this.position, 1)
        val nextLayer0 = level.getBlock(next.position, 0)
        val nextLayer1 = level.getBlock(next.position, 1)

        blockFace = direction

        level.setBlock(block.position, this, direct = true, update = true)
        if (next is BlockLiquid && next.usesWaterLogging()) {
            level.setBlock(next.position, 1, next, direct = true, update = false)
        }

        val head = clone() as BlockBed
        head.isHeadPiece = true
        level.setBlock(next.position, head, direct = true, update = true)

        var thisBed: BlockEntityBed? = null
        try {
            thisBed = createBlockEntity(CompoundTag().putByte("color", item.damage))
            val nextBlock = next.levelBlock as BlockEntityHolder<*>?
            nextBlock!!.createBlockEntity(CompoundTag().putByte("color", item.damage))
        } catch (e: Exception) {
            log.warn(
                "Failed to create the block entity {} at {} and {}",
                getBlockEntityType(), locator, next.locator, e
            )
            thisBed?.close()
            level.setBlock(thisLayer0!!.position, 0, thisLayer0, true)
            level.setBlock(thisLayer1!!.position, 1, thisLayer1, true)
            level.setBlock(nextLayer0!!.position, 0, nextLayer0, true)
            level.setBlock(nextLayer1!!.position, 1, nextLayer1, true)
            return false
        }
        return true
    }

    override fun onBreak(item: Item?): Boolean {
        val direction = blockFace
        if (isHeadPiece) { //This is the Top part of bed
            val other = getSide(direction!!.getOpposite()!!)
            if (other!!.id == id && !(other as BlockBed).isHeadPiece && direction == other.blockFace) {
                level.setBlock(other.position, get(BlockID.AIR), direct = true, update = true)
            }
        } else { //Bottom Part of Bed
            val other = getSide(direction!!)
            if (other!!.id == id && (other as BlockBed).isHeadPiece && direction == other.blockFace) {
                level.setBlock(other.position, get(BlockID.AIR), direct = true, update = true)
            }
        }

        level.setBlock(
            this.position,
            get(BlockID.AIR),
            direct = true,
            update = false
        ) // Do not update both parts to prevent duplication bug if there is two fallable blocks top of the bed

        return true
    }

    override fun toItem(): Item {
        return ItemBed(dyeColor.woolData)
    }

    val dyeColor: DyeColor
        get() {
            val blockEntity = blockEntity
            if (blockEntity != null) {
                return blockEntity.dyeColor
            }

            return DyeColor.WHITE
        }

    override var blockFace: BlockFace
        get() = fromHorizontalIndex(getPropertyValue<Int, IntPropertyType>(CommonBlockProperties.DIRECTION))
        set(face) {
            setPropertyValue<Int, IntPropertyType>(
                CommonBlockProperties.DIRECTION,
                face!!.horizontalIndex
            )
        }

    var isHeadPiece: Boolean
        get() = getPropertyValue<Boolean, BooleanPropertyType>(CommonBlockProperties.HEAD_PIECE_BIT)
        set(headPiece) {
            setPropertyValue<Boolean, BooleanPropertyType>(
                CommonBlockProperties.HEAD_PIECE_BIT,
                headPiece
            )
        }

    var isOccupied: Boolean
        get() = getPropertyValue<Boolean, BooleanPropertyType>(CommonBlockProperties.OCCUPIED_BIT)
        set(occupied) {
            setPropertyValue<Boolean, BooleanPropertyType>(
                CommonBlockProperties.OCCUPIED_BIT,
                occupied
            )
        }

    override fun breaksWhenMoved(): Boolean {
        return true
    }

    override fun sticksToPiston(): Boolean {
        return false
    }

    val isBedValid: Boolean
        get() {
            val dir = blockFace
            val head: Block?
            val foot: Block?
            if (isHeadPiece) {
                head = this
                foot = getSide(dir!!.getOpposite()!!)
            } else {
                head = getSide(dir!!)
                foot = this
            }

            return head!!.id == foot!!.id
                    && (head as BlockBed).isHeadPiece && head.blockFace == dir
                    && !(foot as BlockBed).isHeadPiece && foot.blockFace == dir
        }

    val headPart: BlockBed?
        get() {
            if (isHeadPiece) {
                return this
            }
            val dir = blockFace
            val head = getSide(dir!!)
            if (head!!.id == id && (head as BlockBed).isHeadPiece && head.blockFace == dir) {
                return head
            }
            return null
        }

    val footPart: BlockBed?
        get() {
            if (!isHeadPiece) {
                return this
            }

            val dir = blockFace
            val foot = getSide(dir!!.getOpposite()!!)
            if (foot!!.id == id && !(foot as BlockBed).isHeadPiece && foot.blockFace == dir) {
                return foot
            }
            return null
        }

    override val properties: BlockProperties
        get() = Companion.properties

    companion object {
        val properties: BlockProperties = BlockProperties(
            BlockID.BED,
            CommonBlockProperties.DIRECTION,
            CommonBlockProperties.HEAD_PIECE_BIT,
            CommonBlockProperties.OCCUPIED_BIT
        )

    }
}
